clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -253,6 +282,50 @@ public void forceClose() {
         initiateClose();
     }
 
+    private ClientResponse sendAndAwaitInitPidRequest(Node node) throws IOException {
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, null);
+        return NetworkClientUtils.sendAndReceive(client, request, time);
+    }
+
+    private Node getReadyNode(long remainingTimeMs) throws IOException {
+        Node node = client.leastLoadedNode(time.milliseconds());
+        if (NetworkClientUtils.awaitReady(client, node, time, remainingTimeMs)) {
+            return node;
+        }
+        return null;
+    }
+
+    private void maybeWaitForPid() {
+        if (transactionState == null)
+            return;
+
+        while (!transactionState.hasPid()) {
+            try {
+                Node node = getReadyNode(requestTimeout);
+                if (node != null) {
+                    ClientResponse response = sendAndAwaitInitPidRequest(node);
```

Comment:
Hmm, in sendAndAwaitInitPidRequest(), we just initiate a pid request w/o checking if we actually can send to the node. Not sure if this is safe. NetworkClient also sends internal metadata request. So, it's possible that an ongoing metadata request is still pending on the same node and the send of pid request will hit an IllegalStateException in NetworkClient.doSend().